src/c++/perf_analyzer/test_dataloader.cc

```diff
@@ -0,0 +1,576 @@
+//    { "INPUT1": [3000, 4000] }
+//  ]})"};
+//
+//
+//  MockDataLoader dataloader;
+//  std::shared_ptr<ModelTensorMap> inputs = std::make_shared<ModelTensorMap>();
+//  std::shared_ptr<ModelTensorMap> outputs =
+//  std::make_shared<ModelTensorMap>();
+//
+//  ModelTensor input1 = TestDataLoader::GetTensor("INPUT1");
+//  ModelTensor output1 = TestDataLoader::GetTensor("OUTPUT1");
+//
+//  inputs->insert(std::make_pair(input1.name_, input1));
+//  outputs->insert(std::make_pair(output1.name_, output1));
+//
+//  cb::Error status = dataloader.ReadDataFromJSON(inputs, outputs, json_str1);
+//  CHECK(status.IsOk());
+//  CHECK_EQ(dataloader.GetDataStreamsCount(), 1);
+//  CHECK_EQ(dataloader.GetTotalSteps(0), 2);
+//
+//  status = dataloader.ReadDataFromJSON(inputs, outputs, json_str2);
+//  CHECK(status.IsOk());
+//  CHECK_EQ(dataloader.GetDataStreamsCount(), 3);
+//  CHECK_EQ(dataloader.GetTotalSteps(0), 2);
+//  CHECK_EQ(dataloader.GetTotalSteps(1), 1);
+//  CHECK_EQ(dataloader.GetTotalSteps(2), 1);
+//
+//  status = dataloader.ReadDataFromJSON(inputs, outputs, json_str3);
+//  CHECK(status.IsOk());
+//  CHECK_EQ(dataloader.GetDataStreamsCount(), 4);
+//  CHECK_EQ(dataloader.GetTotalSteps(0), 2);
+//  CHECK_EQ(dataloader.GetTotalSteps(1), 1);
+//  CHECK_EQ(dataloader.GetTotalSteps(2), 1);
+//  CHECK_EQ(dataloader.GetTotalSteps(3), 3);
+//}
+
+TEST_CASE(
+    "dataloader: GenerateData: Is Shape Tensor" *
+    doctest::description("It is illegal to generate data for any Tensor with "
+                         "is_shape_tensor=True"))
+{
+  MockDataLoader dataloader;
+  std::shared_ptr<ModelTensorMap> inputs = std::make_shared<ModelTensorMap>();
+  std::shared_ptr<ModelTensorMap> outputs = std::make_shared<ModelTensorMap>();
+
+  ModelTensor input1 = TestDataLoader::GetTensor("INPUT1");
+  input1.is_shape_tensor_ = true;
+  inputs->insert(std::make_pair(input1.name_, input1));
+
+  bool zero_input = true;
+  size_t string_length = 5;
+  std::string string_data = "FOOBAR";
+  cb::Error status =
+      dataloader.GenerateData(inputs, zero_input, string_length, string_data);
+  CHECK_FALSE(status.IsOk());
+  CHECK_EQ(
+      status.Message(),
+      "can not generate data for shape tensor 'INPUT1', user-provided data is "
+      "needed.");
+}
+
+
+TEST_CASE(
+    "dataloader: GenerateData: Zero Input" *
+    doctest::description(
+        "Calling GenerateData with the zero_input flag set to true should "
+        "result in a single stream with one step with all data having the "
+        "value of 0"))
+{
+  MockDataLoader dataloader;
+  std::shared_ptr<ModelTensorMap> inputs = std::make_shared<ModelTensorMap>();
+  std::shared_ptr<ModelTensorMap> outputs = std::make_shared<ModelTensorMap>();
+
+  ModelTensor input1 = TestDataLoader::GetTensor("INPUT1");
+  input1.shape_ = {3};
+  inputs->insert(std::make_pair(input1.name_, input1));
+
+  bool zero_input = true;
+  size_t string_length = 5;
+  std::string string_data = "FOOBAR";
```

Comment:
Is string_length and string_data used? If not, can they be set to something more non-implicative-of-effect like 0 and empty string?

Reply:
In general, I think GenerateData() needs to be rewritten so that it's not taking unnecessary inputs.

Reply:
I want them to be non-zero so that IF there was a bug it would actually do something wrong (and it would be easier to triage). Imagine if I put 0 for string length, and string length was accidentally being used instead of `zero_input`, but because it was 0 we end up with the correct answer anyways.

And yes, this is all necessary because there are inputs that aren't used.  I have to make the unit testing support the existing code structure before I can consider any refactoring and cleanup.